File: go-backend-interview-prep/03_functions_and_methods/06_defer_deep_dive/main.go
package main

import "fmt"

func main() {
	// --- Argument evaluation timing ---
	fmt.Println("--- Arg evaluation ---")
	x := 10
	defer fmt.Println("deferred x:", x) // captures x=10 NOW
	x = 999
	fmt.Println("current x:", x)
	// Output: current x: 999, then deferred x: 10

	// --- Closure captures current value at EXIT ---
	fmt.Println("\n--- Closure vs direct arg ---")
	y := "hello"
	defer func() { fmt.Println("closure y:", y) }() // captures reference
	defer fmt.Println("direct y:", y)                // captures value NOW
	y = "goodbye"
	// direct prints "hello", closure prints "goodbye"

	// --- Defer in loop: THE TRAP ---
	fmt.Println("\n--- Defer in loop (trap) ---")
	deferInLoopTrap()

	// --- Defer in loop: THE FIX ---
	fmt.Println("\n--- Defer in loop (fix) ---")
	deferInLoopFix()

	// --- Named return + defer ---
	fmt.Println("\n--- Named return + defer ---")
	result := enrichedReturn()
	fmt.Println("result:", result)
}

func deferInLoopTrap() {
	// BAD: all defers pile up until function exits
	for i := 0; i < 3; i++ {
		// Imagine: f, _ := os.Open(files[i])
		// defer f.Close()  // NOT released until function returns!
		defer fmt.Printf("  loop defer (all at end): %d\n", i)
	}
	fmt.Println("  after loop (defers haven't run yet)")
}

func deferInLoopFix() {
	// GOOD: extract body into helper so defer runs per iteration
	for i := 0; i < 3; i++ {
		processItem(i)
	}
}

func processItem(i int) {
	defer fmt.Printf("  defer runs immediately after iteration %d\n", i)
	fmt.Printf("  processing %d\n", i)
}

// Named return values can be modified by deferred closures.
func enrichedReturn() (result string) {
	result = "original"
	defer func() {
		result = result + " (enriched by defer)"
	}()
	return result
}

File: go-backend-interview-prep/03_functions_and_methods/06_defer_deep_dive/main.py
# Python equivalent of Defer Deep Dive (compare with main.go)
# Python has no defer. Use try/finally or context managers.


def main():
    # --- try/finally = basic defer ---
    print("--- try/finally ---")
    x = 10
    try:
        x = 999
        print("current x:", x)
    finally:
        # Unlike Go defer, this sees the CURRENT value of x
        print("finally x:", x)  # 999 (Go defer would print 10)

    # --- Context manager (Pythonic defer for resources) ---
    print("\n--- Context manager ---")
    # with open("file.txt") as f:
    #     data = f.read()
    # # f is automatically closed here

    # --- Loop cleanup: context manager per iteration ---
    print("\n--- Loop cleanup ---")
    for i in range(3):
        process_item(i)  # each call has its own try/finally

    # --- No named return + defer pattern in Python ---
    print("\n--- Return enrichment ---")
    # Python has no direct equivalent of Go's named return + defer.
    # Use a wrapper or try/except/finally instead.
    result = enriched_return()
    print("result:", result)

    # --- Key difference ---
    print("\nKey: Go defer captures arg values at defer-time.")
    print("     Python finally sees current values at finally-time.")


def process_item(i: int) -> None:
    try:
        print(f"  processing {i}")
    finally:
        print(f"  cleanup after iteration {i}")


def enriched_return() -> str:
    """Closest equivalent to Go named return + defer."""
    result = "original"
    try:
        return result
    finally:
        # In Python, finally runs but cannot change the return value
        # (unlike Go deferred closures modifying named returns)
        pass
    # To modify return, restructure the code differently


if __name__ == "__main__":
    main()
